validator/client/src/main/java/tech/pegasys/artemis/validator/client/DutyScheduler.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.validator.client;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.artemis.util.async.AsyncRunner;
+import tech.pegasys.artemis.util.async.SafeFuture;
+import tech.pegasys.artemis.util.bls.BLSPublicKey;
+import tech.pegasys.artemis.validator.api.Duty;
+import tech.pegasys.artemis.validator.api.ValidatorApiChannel;
+import tech.pegasys.artemis.validator.api.ValidatorDuties;
+import tech.pegasys.artemis.validator.api.ValidatorTimingChannel;
+import tech.pegasys.artemis.validator.client.duties.AttestationProductionDuty;
+import tech.pegasys.artemis.validator.client.duties.ValidatorDutyFactory;
+
+public class DutyScheduler implements ValidatorTimingChannel {
+  private static final Logger LOG = LogManager.getLogger();
+  final AtomicReference<UnsignedLong> latestScheduledEpoch = new AtomicReference<>();
+  private final ConcurrentMap<UnsignedLong, Duty> blockProposalDuties = new ConcurrentHashMap<>();
```

Comment:
It might be a good idea to rename Duty to BlockProductionDuty, so that's it's more in line with the AttestationProductionDuty naming.